### PR TITLE
Basic Permissions system and Server Game Mode!

### PR DESCRIPTION
**Pull request by killzoms**
Finally got around to adding a basic permissions system and server game mode option!

### CODE REVIEW

File: NitroxServer/Serialization/ServerConfig.cs [hunk 1]
```diff
@@ -41,6 +43,50 @@ public int MaxConnections
...
+                }
+                return _gameMode ?? DEFAULT_GAMEMODE;
+            }
+        }
+
+        GameModeOption ParseGameMode(string stringGameMode)
+        {
+            GameModeOption gameMode = GameModeOption.Survival;
+            stringGameMode = stringGameMode.ToLower(); // Lets be frank people have habits
+            switch (stringGameMode)
```
Comment (Sunrunner37) [inline NitroxServer/Serialization/ServerConfig.cs]: Might be cleaner as a dictionary above:

```
Dictionary<string, GameModeOption> gameModeByConfigKey = new Dictionary<string, GameModeOption> {
  {"survival", GameModeOption.Survival},
  {"creative", GameModeOption.Creative},
  {"hardcore", GameModeOption.Hardcore},
  {"permadeath, GameModeOption.Survival | GameModeOption.Permadeath}
};
```

Then you can remove this method and simply use the map in the `GameMode` getter

File: NitroxServer/Communication/Packets/Processors/ServerCommandProcessor.cs [hunk 2]
```diff
@@ -1,23 +1,34 @@
...
             this.cmdProcessor = cmdProcessor;
+            this.playerData = playerData;
         }
 
         public override void Process(ServerCommand packet, Player player)
         {
             string msg = string.Join(" ", packet.CmdArgs);
 
-            cmdProcessor.ProcessCommand(msg);
+            if (playerData.GetPermissions(player.Name) >= Perms.Admin)
```
Comment (Sunrunner37) [inline NitroxServer/Communication/Packets/Processors/ServerCommandProcessor.cs]: Maybe this should be handled outside of the actual command?  Commands can have a field like `MinPermissionRequired` and we don't even invoke if they don't meet the requirement.  That way you don't have to copy/paste this packet sending into every command.

File: NitroxServer/GameLogic/Players/PlayerData.cs [hunk 3]
```diff
@@ -194,6 +212,9 @@ public class PersistedPlayerData
             [ProtoMember(6)]
             public string SubRootGuid { get; set; }
 
+            [ProtoMember(7)]
+            public Perms Permissions { get; set; }
```
Comment (Sunrunner37) [inline NitroxServer/GameLogic/Players/PlayerData.cs]: Is this nullable?  Should we assign a default value if the serializer doesn't have one from the file (backwards compatibility)

File: NitroxServer/ConsoleCommands/OpCommand.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        public override void RunCommand(string[] args)
+        {
+            playerData.UpdatePlayerPermissions(args[0], Perms.Admin);
+        }
+
+        public override bool VerifyArgs(string[] args)
+        {
+            bool playerFound = false;
+            foreach(Player player in playerManager.GetPlayers())
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/OpCommand.cs]: Maybe add an `Optional<Player> GetPlayerByName()` in player manager to make this cleaner.

File: NitroxServer/Serialization/ServerConfig.cs (hunk 1 above)
Comment (killzoms) [reply NitroxServer/Serialization/ServerConfig.cs]: used TryGetValue after creating Dictionary

File: NitroxServer/Communication/Packets/Processors/ServerCommandProcessor.cs (hunk 2 above)
Comment (killzoms) [reply NitroxServer/Communication/Packets/Processors/ServerCommandProcessor.cs]: checked off the list

File: NitroxServer/GameLogic/PlayerManager.cs [hunk 5]
```diff
@@ -123,6 +123,30 @@ public void PlayerDisconnected(Connection connection)
...
+        public bool TryGetPlayerByName(string playerName, out Player foundPlayer)
+        {
+            lock (assetsByConnection)
+            {
+                foundPlayer = null;
+                foreach (Player player in ConnectedPlayers())
+                {
+                    if (player.Name == playerName)
+                    {
+                        foundPlayer = player;
```
Comment (Sunrunner37) [inline NitroxServer/GameLogic/PlayerManager.cs]: minor: return true after this line as an optimization.

File: NitroxServer/App.config [hunk 6]
```diff
@@ -7,5 +7,7 @@
         <add key="MaxConnections" value="100"/>
         <add key="DefaultPortNumber" value="11000"/>
         <add key="SaveInterval" value="60000"/>
+        <add key="GameMode" value="Survival"/>
+        <add key="DisableConsole" value="true"/>
```
Comment (Sunrunner37) [inline NitroxServer/App.config]: Maybe default to false while we are still in development.

File: NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs [hunk 7]
```diff
@@ -54,6 +59,39 @@ public void ProcessCommand(string msg)
...
+
+            if (cmd.RequiredPermLevel > perms)
+            {
+                player.SendPacket(new ChatMessage(ChatMessage.SERVER_ID, "You do not have permission for this command!"));
+                return;
+            }
+
+            if (cmd.SupportsClientSide)
+            {
+                RunPlayerCommand(cmd, parts, player);
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs]: Do we have the use-cases to support an entirely separate method signature?  I think a command like `/give <item> <player>` would always have player as parameter.  And if it always assumed the executing player you would have to have two different commands in this model. Thoughts?

File: NitroxServer/ConsoleCommands/SayCommand.cs [hunk 8]
```diff
@@ -3,22 +3,28 @@
...
+using NitroxModel.DataStructures.GameLogic;
 
 namespace NitroxServer.ConsoleCommands
 {
     internal class SayCommand : Command
     {
         private readonly PlayerManager playerManager;
 
-        public SayCommand(PlayerManager playerManager) : base("say", Optional<string>.Of("<message>"), "say Even the lowliest of cogs needs to say something SO SAY SOMETHING!", new[] {"broadcast"})
+        public SayCommand(PlayerManager playerManager) : base("say", NitroxModel.DataStructures.GameLogic.Perms.Admin, Optional<string>.Of("<message>"), "say Even the lowliest of cogs needs to say something SO SAY SOMETHING!", new[] {"broadcast"})
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/SayCommand.cs]: minor: why fully qualified name?

File: NitroxServer/GameLogic/PlayerManager.cs [hunk 9]
```diff
@@ -123,6 +123,30 @@ public void PlayerDisconnected(Connection connection)
             }
         }
 
+
+        /// <summary>
+        /// Trys to return a player by name
+        /// </summary>
+        /// <param name="playerName">The player to find</param>
+        /// <param name="foundPlayer">Will be null if return value is false</param>
+        /// <returns>true if the player was found false otherwise</returns>
```
Comment (Sunrunner37) [inline NitroxServer/GameLogic/PlayerManager.cs]: minor: commenting doesn't add value.  Summary is the name of the method, parameters are self explainable, and the usage aligns with c#'s Try* that is used everywhere in the code.

File: NitroxServer/Serialization/ServerConfig.cs [hunk 10]
```diff
@@ -41,6 +55,42 @@ public int MaxConnections
...
+            {
+                if (_gameMode == null && ConfigurationManager.AppSettings[GAMEMODE_SETTING] != null)
+                {
+                    _gameMode = ParseGameMode(ConfigurationManager.AppSettings[GAMEMODE_SETTING]);
+                }
+                return _gameMode ?? DEFAULT_GAMEMODE;
+            }
+        }
+
+        GameModeOption ParseGameMode(string stringGameMode)
```
Comment (Sunrunner37) [inline NitroxServer/Serialization/ServerConfig.cs]: minor: explicit accessor `private` per style.

File: NitroxServer/ConsoleCommands/SayCommand.cs [hunk 11]
```diff
@@ -3,22 +3,28 @@
...
         {
             this.playerManager = playerManager;
+            SupportsClientSide = true;
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/SayCommand.cs]: Are some not supported client side?

File: NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs [hunk 12]
```diff
@@ -66,5 +104,18 @@ private void RunCommand(Command command, string[] parts)
...
         }
+
+        private void RunPlayerCommand(Command command, string[] parts, Player player)
+        {
+            string[] args = parts.Skip(1).ToArray();
+
+            if (command.VerifyArgs(args))
+            {
+                command.RunCommand(args, player);
+                return;
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs]: minor: in this circumstance, i think `else` is more intuitive than `return`

File: NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs [hunk 13]
```diff
@@ -54,6 +59,39 @@ public void ProcessCommand(string msg)
...
+            }
+
+            string[] parts = msg.Split()
+                .Where(arg => !string.IsNullOrEmpty(arg))
+                .ToArray();
+
+            Command cmd;
+            if (!commands.TryGetValue(parts[0], out cmd))
+            {
+                return;
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs]: Should we send a packet saying the command is invalid?

File: NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs (hunk 7 above)
Comment (killzoms) [reply NitroxServer/ConsoleCommands/Processor/ConsoleCommandProcessor.cs]: Though i used the different methods due to sending info back to the player without creating a new method in PlayerManager to try getting the player based on the player name

File: NitroxServer/ConsoleCommands/SayCommand.cs (hunk 11 above)
Comment (killzoms) [reply NitroxServer/ConsoleCommands/SayCommand.cs]: used the SupportsClientSide to differentiate between commands like exit and say, there isnt a need to tell the client about the exit command so we dont have to differentiate between the Player execution method and the server execution method

**Comment by killzoms** [comment]
should be a new command added with the latest batch of commits "/w <name> <msg>" Everyones favorite whisper

File: NitroxServer/Player.cs [hunk 14]
```diff
@@ -71,7 +72,21 @@ public bool CanSee(Entity entity)
 
         public void SendPacket(Packet packet)
         {
-            connection.SendPacket(packet);            
+            if (Id == ChatMessage.SERVER_ID) // this is probably just a small hack
```
Comment (Sunrunner37) [inline NitroxServer/Player.cs]: Can we put this hack in the caller instead?

File: NitroxServer/ConsoleCommands/WhisperCommand.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+            }
+            else
+            {
+                if (player.Id != ChatMessage.SERVER_ID)
+                {
+                    player.SendPacket(new ChatMessage(ChatMessage.SERVER_ID, "Player not found!"));
+                }
+                else
+                {
+                    Log.Info("Player not found!");
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/WhisperCommand.cs]: We don't want to log an error as this is normal operation (someone running the command on the server)

File: NitroxServer/ConsoleCommands/WhisperCommand.cs [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    class WhisperCommand : Command
+    {
+        private readonly PlayerManager playerManager;
+
+        public WhisperCommand(PlayerManager playerManager) : base("w", Perms.Player, Optional<string>.Of("w <PlayerName> <msg>"))
+        {
+            this.playerManager = playerManager;
+        }
+
+        public override void RunCommand(string[] args, Player player)
```
Comment (Sunrunner37) [inline NitroxServer/ConsoleCommands/WhisperCommand.cs]: Let's pass an Optional<Player> into these and we can do Optional<Player>.empty() if it is the server.  Then we don't have to check for the server id.

File: NitroxServer/ConsoleCommands/WhisperCommand.cs (hunk 15 above)
Comment (killzoms) [reply NitroxServer/ConsoleCommands/WhisperCommand.cs]: the reason for the player not found being logged is in the event someone mistypes a players name

**Pull request merged**